man/man1/reservoir2cdf.1
.TH reservoir2cdf 1  "USER COMMANDS"
.SH NAME
reservoir2cdf \- convert CLM reservoir file into NetCDF file
.SH SYNOPSIS
.B reservoir2cdf
[\-h] [\-v] [\-global] [\-longheader] [\-cellsize size] [[\-attr name=value] ...] [\-metafile] [\-raw] [\-netcdf4] [\-compress level]
.I [gridfile] reservoirfile netcdffile
.SH DESCRIPTION
Program converts CLM reservoir input data into NetCDF input data.
.SH OPTIONS
.TP
\-h,\--help
print help text.
.TP
\-v,\--version
print LPJmL version.
.TP
\-global
use global grid for NetCDF file.
.TP
\-longheader
force version of CLM file to 2.
.TP
\-cellsize size
set cell size of grid file to size.
.TP
\-attr name=value
set global attribute name in output files to value.
.TP
\-metafile
set the input format to JSON metafile instead of CLM.
.TP
\-raw
set the input format to raw instead of CLM.
.TP
\-netcdf4
File written is in NetCDF4 format, compress flag is enabled.
.TP
\-compress level
compression level for NetCDF4 files.
.TP
.I gridfile
filename of grid data file. If not specified grid name is read from metafile.
.TP
.I reservoirfile
filename of CLM reservoir file with 10 bands.
.TP
.I netcdffile
filename of NetCDF file created.
.SH EXAMPLES
.TP
Create reservoir data file:
.B reservoir2cdf
grid.clm reservoir.clm reservoir.nc
.TP
Create data file using JSON metafile:
.B reservoir2cdf
-metafile reservoir.clm.json reservoir.nc
.PP
.SH EXIT STATUS
.B reservoir2cdf
returns a zero exit status if the NetCDF file is created.
Non zero is returned in case of failure.

.SH AUTHORS

See AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
clm2cdf(1), bin2cdf(1), country2cdf(1), drainage2cdf(1), regridlpj(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


## [5.9.24] - 2025-05-15

### Contributors

- author: David Hötten (davidho@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Maik Billing (billing@pik-potsdam.de)

### Added

- Merge request template ```.gitlab/merge_request_templates/default.md``` that includes several checklists

## [5.9.23] - 2025-05-08

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Jens Heinke (heinke@pik-potsdam.de), Fabian Stenzel(stenzel@pik-potsdam.de)

### Changed

- Option `-check` of `configure.sh` enables now run-time checks of memory leaks and memory access out of bounds and undefined variables for gcc and icx compiler. Optimization is not disabled. Use options `-debug -check` to disable.
- Datatype `List` used in `fprintfiles.c`.
- All allocated memory is now freed in utilities.
- Prescribed landcover input file changed in `input.cjson` to a file with the same number of natural PFTs.
- All remaining unsafe calls to `sprintf()` replaced by calls to `snprinf()` or `getsprinf()`.

### Fixed

- Check for missing values for integer input corrected in `readclimate_netcdf.c`.
- Memory leak fixed in `filesexist.c`.
- Calculation of global area fixed in `printglobal.c`.
- Check for null attribute pointer added in `fprintjson.c`.
- Uninitialized grid name initialized in `mathclm.c`.
- Call to `snprintf()` replaced by `getsprintf()` in `reservoir2cdf.c`.
- Datatype for index and writing the header corrected in `lpjcat.c`.
- Functions for qsort(), bisect(), and leftmostzero() changed to avoid run-time errors with `-check` option.
- Pointer set to NULL  if `k_est`is NULL in `initmanage.c`.
- Missing argument added in opening aquifer file for IMAGE.
- Scaling of coordinates fixed in `joingrid.c`.
- Fraction pointer initialized to NULL in `freadresdata.c` to avoid SEGV in `lpjprint.c` if reservoir data is read from restart file.
- Index set correctly for outflow cells in `regriddrain.c`.
- Check for maximum discharge length corrected in `printdrain.c`.
- Number of years of clm file set to 1 in `grid2clm.c`.
- Function `fscanlandcovermap.c` corrected (issue https://github.com/PIK-LPJmL/LPJmL/issues/48).
- Access out of bounds for `soil->freezdepth` fixed in `soiltemp.c`.
- Format specifier changed in `cat2bsq.c` and `cutclm.c` to compile without warnings using the clang compiler (issue #372).


## [5.9.22] - 2025-03-14

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Hester Biemans (hester.biemans@wur.nl)
- code review: Susanne Rolinski (rolinski@pik-potsdam.de), Marie Hemmen (hemmen@pik-potsdam.de)

### Changed

- `input_netcdf.cjson` changed to the default dataset of `input.cjson` but in NetCDF format.
- Code changed to compile under Windows OS.
- New function `getsprintf()` added to allocate and print formatted output into string. Function replaces call to `snprintf()` function.
- Default settings for `GIT_HASH` and `GIT_REPO` added.
- Duplicate filenames removed from list of input/output filenames in utility `lpjfiles`.

### Added

- Option `-int` added to utility `cdf2clm`.
- Option `-latlon` added to utility `cdf2coord` in order to change the order of the CLM grid file.
- Utility `cdf2reservoir` added to convert NetCDF reservoir file into CLM file.
- Utility `reservoir2cdf` added to convert CLM reservoir file into a NetCDF file using the soil code NetCDF file.
- Macro `NETCDF_INPUT` added in `lpjml_config.cjson` to enable NetCDF input.
- Reservoir, irrigation neighbor and drainage data can now be in NetCDF format:
```java
"drainage" :           { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/drainage.nc"},
"neighb_irrig" :       { "fmt" : "cdf", "var" : "index", "name" : "cru_netcdf/neighb_irrig.nc"},
"river" :              { "fmt" : "cdf", "var" : "riverlen", "name" : "cru_netcdf/drainage.nc"},
"reservoir" :          { "fmt" : "cdf", "var" : "year", "name" : "cru_netcdf/reservoir.nc"},
"capacity_reservoir" : { "fmt" : "cdf", "var" : "capacity", "name" : "cru_netcdf/reservoir.nc"},

src/utils/reservoir2cdf.c
/**************************************************************************************/
/**                                                                                \n**/
/**                r  e  s  e  r  v  o  i  r  2  c  d  f  .  c                     \n**/
/**                                                                                \n**/
/**     Converts CLM reservoir file into NetCDF data file                          \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#ifdef USE_NETCDF
#include <netcdf.h>

#define error(rc) if(rc) {free(lon);free(lat);fprintf(stderr,"ERROR427: Cannot write '%s': %s.\n",filename,nc_strerror(rc)); nc_close(cdf->ncid); free(cdf);return NULL;}

#define USAGE "Usage: %s [-h] [-v] [-metafile] [-longheader] [-global] [-cellsize size]\n       [[-attr name=value] ...] [-raw] [-netcdf4] [-compress level]\n       [gridfile] clmfile netcdffile\n"

typedef struct
{
  Coord_array *index;
  int ncid;
  int varid_year;
  int varid_area;
  int varid_capacity;
  int varid_inst_cap;
  int varid_height;
  int varid_purpose;
} Cdf;

static Cdf *create_cdf(const char *filename,
                       const char *source,
                       const char *history,
                       float miss,
                       int imiss,
                       const char *args,
                       const Attr *global_attrs,
                       int n_global,
                       const Header *header,
                       int compress,
                       Bool isnetcdf4,
                       Coord_array *array)
{
  Cdf *cdf;
  double *lon,*lat;
  int i,rc,dim[3];
  char *s;
  time_t t;
  int lat_var_id,lon_var_id,lat_dim_id,lon_dim_id;
  int purpose_dim_id;
  cdf=new(Cdf);
  lon=newvec(double,array->nlon);
  if(lon==NULL)
  {
    printallocerr("lon");
    free(cdf);
    return NULL;
  }
  lat=newvec(double,array->nlat);
  if(lat==NULL)
  {
    printallocerr("lat");
    free(lon);
    free(cdf);
    return NULL;
  }
  for(i=0;i<array->nlon;i++)
    lon[i]=array->lon_min+i*header->cellsize_lon;
  for(i=0;i<array->nlat;i++)
    lat[i]=array->lat_min+i*header->cellsize_lat;
  cdf->index=array;
  rc=nc_create(filename,(isnetcdf4) ? NC_CLOBBER|NC_NETCDF4 : NC_CLOBBER,&cdf->ncid);
  if(rc)
  {
    fprintf(stderr,"ERROR426: Cannot create file '%s': %s.\n",
            filename,nc_strerror(rc));
    free(lon);
    free(lat);
    free(cdf);
    return NULL;
  }
  rc=nc_def_dim(cdf->ncid,LAT_DIM_NAME,array->nlat,&lat_dim_id);
  error(rc);
  rc=nc_def_dim(cdf->ncid,LON_DIM_NAME,array->nlon,&lon_dim_id);
  error(rc);
  if(source!=NULL)
  {
    rc=nc_put_att_text(cdf->ncid,NC_GLOBAL,"source",strlen(source),source);
    error(rc);
  }
  time(&t);
  if(history!=NULL)
  {
    s=getsprintf("%s\n%s: %s",history,strdate(&t),args);
  }
  else
  {
    s=getsprintf("%s: %s",strdate(&t),args);
  }
  check(s);
  rc=nc_put_att_text(cdf->ncid,NC_GLOBAL,"history",strlen(s),s);
  free(s);
  error(rc);
  for(i=0;i<n_global;i++)
  {
    rc=nc_put_att_text(cdf->ncid,NC_GLOBAL,global_attrs[i].name,strlen(global_attrs[i].value),global_attrs[i].value);
    error(rc);
  }
  rc=nc_def_var(cdf->ncid,LAT_NAME,NC_DOUBLE,1,&lat_dim_id,&lat_var_id);
  error(rc);
  rc=nc_def_var(cdf->ncid,LON_NAME,NC_DOUBLE,1,&lon_dim_id,&lon_var_id);
  error(rc);
  rc=nc_put_att_text(cdf->ncid,lon_var_id,"units",strlen("degrees_east"),
                     "degrees_east");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"long_name",strlen(LON_LONG_NAME),LON_LONG_NAME);
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"standard_name",strlen(LON_STANDARD_NAME),LON_STANDARD_NAME);
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lon_var_id,"axis",strlen("X"),"X");
  error(rc);
  rc=nc_put_att_text(cdf->ncid,lat_var_id,"units",strlen("degrees_north"),
                   "degrees_north");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"long_name",strlen(LAT_LONG_NAME),LAT_LONG_NAME);
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"standard_name",strlen(LAT_STANDARD_NAME),LAT_STANDARD_NAME);
  error(rc);
  rc=nc_put_att_text(cdf->ncid, lat_var_id,"axis",strlen("Y"),"Y");
  error(rc);
  rc=nc_def_dim(cdf->ncid,"purpose",NPURPOSE,&purpose_dim_id);
  error(rc);
  dim[0]=lat_dim_id;
  dim[1]=lon_dim_id;
  rc=nc_def_var(cdf->ncid,"year",NC_INT,2,dim,&cdf->varid_year);
  rc=nc_put_att_text(cdf->ncid, cdf->varid_year,"standard_name",strlen("year"),"year");
  nc_put_att_int(cdf->ncid, cdf->varid_year,"missing_value",NC_INT,1,&imiss);
  rc=nc_put_att_int(cdf->ncid, cdf->varid_year,"_FillValue",NC_INT,1,&imiss);
  error(rc);
  rc=nc_def_var(cdf->ncid,"area",NC_FLOAT,2,dim,&cdf->varid_area);
  rc=nc_put_att_text(cdf->ncid, cdf->varid_area,"units",strlen("km2"),"km2");
  nc_put_att_float(cdf->ncid, cdf->varid_area,"missing_value",NC_FLOAT,1,&miss);
  rc=nc_put_att_float(cdf->ncid, cdf->varid_area,"_FillValue",NC_FLOAT,1,&miss);
  error(rc);
  rc=nc_def_var(cdf->ncid,"capacity",NC_FLOAT,2,dim,&cdf->varid_capacity);
  rc=nc_put_att_text(cdf->ncid, cdf->varid_capacity,"units",strlen("km3"),"km3");
  nc_put_att_float(cdf->ncid, cdf->varid_capacity,"missing_value",NC_FLOAT,1,&miss);
  rc=nc_put_att_float(cdf->ncid, cdf->varid_capacity,"_FillValue",NC_FLOAT,1,&miss);
  error(rc);
  rc=nc_def_var(cdf->ncid,"inst_cap",NC_INT,2,dim,&cdf->varid_inst_cap);
  nc_put_att_int(cdf->ncid, cdf->varid_inst_cap,"missing_value",NC_INT,1,&imiss);
  rc=nc_put_att_int(cdf->ncid, cdf->varid_inst_cap,"_FillValue",NC_INT,1,&imiss);
  error(rc);
  rc=nc_def_var(cdf->ncid,"height",NC_INT,2,dim,&cdf->varid_height);
  nc_put_att_int(cdf->ncid, cdf->varid_height,"missing_value",NC_INT,1,&imiss);
  rc=nc_put_att_int(cdf->ncid, cdf->varid_height,"_FillValue",NC_INT,1,&imiss);
  rc=nc_put_att_text(cdf->ncid, cdf->varid_height,"units",strlen("m"),"m");
  rc=nc_put_att_text(cdf->ncid, cdf->varid_height,"standard_name",strlen("height"),"height");
  error(rc);
  rc=nc_put_att_text(cdf->ncid, cdf->varid_height,"long_name",strlen("reservoir height"),"reservoir height");
  error(rc);
  dim[0]=purpose_dim_id;
  dim[1]=lat_dim_id;
  dim[2]=lon_dim_id;
  error(rc);
  rc=nc_def_var(cdf->ncid,"purpose",NC_INT,3,dim,&cdf->varid_purpose);
  error(rc);
  nc_put_att_int(cdf->ncid, cdf->varid_purpose,"missing_value",NC_INT,1,&imiss);
  rc=nc_put_att_int(cdf->ncid, cdf->varid_height,"_FillValue",NC_INT,1,&imiss);
  if(isnetcdf4 && compress)
  {
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_year, 0, 1,compress);
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_area, 0, 1,compress);
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_capacity, 0, 1,compress);
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_inst_cap, 0, 1,compress);
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_height, 0, 1,compress);
    rc=nc_def_var_deflate(cdf->ncid, cdf->varid_purpose, 0, 1,compress);
    error(rc);
  }
  rc=nc_enddef(cdf->ncid);
  error(rc);
  rc=nc_put_var_double(cdf->ncid,lat_var_id,lat);
  error(rc);
  rc=nc_put_var_double(cdf->ncid,lon_var_id,lon);
  error(rc);
  free(lat);
  free(lon);
  return cdf;
} /* of 'create_cdf' */

static Bool write_reservoir_cdf(const Cdf *cdf,const Reservoir r[],int size,float miss,int imiss)
{
  int i,rc,index;
  size_t offsets[3],counts[3];
  float *grid;
  int *igrid;
  grid=newvec(float,cdf->index->nlon*cdf->index->nlat);
  if(grid==NULL)
  {
    printallocerr("grid");
    return TRUE;
  }
  igrid=newvec(int,cdf->index->nlon*cdf->index->nlat);
  if(igrid==NULL)
  {
    free(grid);
    printallocerr("grid");
    return TRUE;
  }
  for(i=0;i<cdf->index->nlon*cdf->index->nlat;i++)
  {
    igrid[i]=imiss;
    grid[i]=miss;
  }
  for(i=0;i<size;i++)
    igrid[cdf->index->index[i]]=r[i].year;
  rc=nc_put_var_int(cdf->ncid,cdf->varid_year,igrid);
  if(rc!=NC_NOERR)
  {
    free(grid);
    free(igrid);
    fprintf(stderr,"ERROR431: Cannot write output data for year: %s.\n",
            nc_strerror(rc));
    return TRUE;
  }
  for(i=0;i<size;i++)
    igrid[cdf->index->index[i]]=r[i].height;
  rc=nc_put_var_int(cdf->ncid,cdf->varid_height,igrid);
  for(i=0;i<size;i++)
    igrid[cdf->index->index[i]]=r[i].inst_cap;
  rc=nc_put_var_int(cdf->ncid,cdf->varid_inst_cap,igrid);
  for(i=0;i<size;i++)
    grid[cdf->index->index[i]]=r[i].area;
  rc=nc_put_var_float(cdf->ncid,cdf->varid_area,grid);
  for(i=0;i<size;i++)
    grid[cdf->index->index[i]]=r[i].capacity;
  rc=nc_put_var_float(cdf->ncid,cdf->varid_capacity,grid);
  offsets[1]=offsets[2]=0;
  counts[0]=1;
  counts[1]=cdf->index->nlat;
  counts[2]=cdf->index->nlon;
  for(index=0;index<NPURPOSE;index++)
  {
    offsets[0]=index;
    for(i=0;i<size;i++)
      igrid[cdf->index->index[i]]=r[i].purpose[index];
    rc=nc_put_vara_int(cdf->ncid,cdf->varid_purpose,offsets,counts,igrid);
    if(rc!=NC_NOERR)
    {
      free(grid);
      free(igrid);
      fprintf(stderr,"ERROR431: Cannot write output data for purpose[%d]: %s.\n",
              index,nc_strerror(rc));
      return TRUE;
    }
  }
  free(grid);
  free(igrid);
  return FALSE;
} /* of 'write_reservoir_cdf' */

static void close_cdf(Cdf *cdf)
{
  freecoordarray(cdf->index);
  nc_close(cdf->ncid);
  free(cdf);
} /* of 'close_cdf' */

#endif
int main(int argc,char **argv)
{
#ifdef USE_NETCDF
  FILE *file;
  Coordfile coordfile;
  Coord_array *index;
  Coord *grid,res;
  Cdf *cdf;
  Header header;
  String headername;
  size_t offset;
  Attr *global_attrs=NULL;
  Attr *global_attrs2=NULL;
  int i,ngrid,version,iarg,compress,setversion;
  Bool swap,isglobal,israw,ismeta,isnetcdf4,rc;
  int n_global,n_global2;
  float cellsize_lon,cellsize_lat;
  Reservoir *rdata;
  char *endptr,*arglist;
  char *pos;
  const char *progname;
  char *grid_filename,*path;
  Filename grid_name;
  char *filename,*outname;
  size_t filesize;
  char *source=NULL,*history=NULL;
  float miss=MISSING_VALUE_FLOAT;
  int imiss=MISSING_VALUE_INT;
  compress=0;
  cellsize_lon=cellsize_lat=0;
  isglobal=FALSE;
  israw=FALSE;
  ismeta=FALSE;
  isnetcdf4=FALSE;
  setversion=READ_VERSION;
  grid_name.fmt=CLM;
  n_global=0;
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("   reservoir2cdf (" __DATE__ ") Help\n"
               "   ================================\n\n"
               "Convert CLM reservoir data into NetCDF input data for LPJmL version %s\n\n",
               getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h,--help        print this help text\n"
               "-v,--version     print LPJmL version\n"
               "-global          use global grid for NetCDF file\n"
               "-longheader      force version of CLM file to 2\n"
               "-cellsize s      set cell size, default is 0.5\n"
               "-metafile        set the input format to JSON metafile instead of CLM\n"
               "-raw             set the input format to raw instead of CLM\n"
               "-netcdf4         file written is in NetCDF4 format\n"
               "-compress l      set compression level for NetCDF4 files\n"
               "-attr name=value set global attribute name to value in NetCDF file\n"
               "gridfile         filename of grid data file\n"
               "clmfile          filename of CLM data file\n"
               "netcdffile       filename of NetCDF file created\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               progname);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[iarg],"-v") || !strcmp(argv[iarg],"--version"))
      {
        puts(getversion());
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[iarg],"-global"))
        isglobal=TRUE;
      else if(!strcmp(argv[iarg],"-longheader"))
        setversion=2;
      else if(!strcmp(argv[iarg],"-raw"))
        israw=TRUE;
      else if(!strcmp(argv[iarg],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[iarg],"-netcdf4"))
        isnetcdf4=TRUE;
      else if(!strcmp(argv[iarg],"-attr"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,
                  "Error: Argument missing for '-attr' option.\n"
                 USAGE,progname);
          return EXIT_FAILURE;
        }
        pos=strchr(argv[++iarg],'=');
        if(pos==NULL)
        {
          fprintf(stderr,
                  "Error: Missing '=' for '-attr' option.\n"
                 USAGE,progname);
          return EXIT_FAILURE;
        }
        *pos='\0';
        global_attrs=realloc(global_attrs,(n_global+1)*sizeof(Attr));
        check(global_attrs);
        global_attrs[n_global].name=strdup(argv[iarg]);
        check(global_attrs[n_global].name);
        global_attrs[n_global].value=strdup(pos+1);
        check(global_attrs[n_global].value);
        n_global++;
      }
      else if(!strcmp(argv[iarg],"-cellsize"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-cellsize'.\n"
                  USAGE,progname);
          return EXIT_FAILURE;
        }
        cellsize_lon=cellsize_lat=(float)strtod(argv[++iarg],&endptr);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-cellsize'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-compress"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Error: Missing argument after option '-compress'.\n"
                  USAGE,progname);
          return EXIT_FAILURE;
        }
        compress=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Error: Invalid number '%s' for option '-compress'.\n",argv[iarg]);
          return EXIT_FAILURE;
        }
        if(compress<0 || compress>9)
        {
          fprintf(stderr,"Error: Invalid compression value %d, must be in [0,9].\n",
                  compress);
          return EXIT_FAILURE;
        }
      }
      else
      {
        fprintf(stderr,"Error: Invalid option '%s'.\n"
                USAGE,argv[iarg],progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(ismeta && argc==iarg+2)
  {
    filename=argv[iarg];
    outname=argv[iarg+1];
  }
  else if(argc<iarg+3)
  {
    fprintf(stderr,"Error: Missing arguments.\n"
            USAGE,progname);
    return EXIT_FAILURE;
  }
  else
  {
    filename=argv[iarg+1];
    outname=argv[iarg+2];
  }
  if(ismeta)
  {
    header.cellsize_lon=header.cellsize_lat=0.5;
    header.firstyear=1901;
    header.firstcell=0;
    header.nyear=1;
    header.nbands=10;
    header.nstep=1;
    header.datatype=LPJ_INT;
    header.order=CELLYEAR;

    file=openmetafile(&header,NULL,NULL,&global_attrs2,&n_global2,&source,&history,NULL,NULL,NULL,NULL,&grid_name,NULL,NULL,&swap,&offset,filename,TRUE);
    if(file==NULL)
      return EXIT_FAILURE;
    if(fseek(file,offset,SEEK_CUR))
    {
      fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",filename,offset);
      fclose(file);
      return EXIT_FAILURE;
    }
    if(global_attrs2!=NULL)
    {
      mergeattrs(&global_attrs,&n_global,global_attrs2,n_global2,FALSE);
      freeattrs(global_attrs2,n_global2);
    }
  }
  else
  {
    file=fopen(filename,"rb");
    if(file==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",filename,strerror(errno));
      return EXIT_FAILURE;
    }
  }
  if(argc!=iarg+2)
  {
    grid_filename=argv[iarg];
  }
  else
  {
    if(grid_name.name==NULL)
    {
      fprintf(stderr,"Error: grid filename must be specified in '%s' metafile.\n",filename);
      return EXIT_FAILURE;
    }
    path=getpath(filename);
    grid_filename=addpath(grid_name.name,path);
    if(grid_filename==NULL)
    {
      printallocerr("name");
      return EXIT_FAILURE;
    }
    free(grid_name.name);
    free(path);
  }
  /* Open and read grid file */
  grid_name.name=grid_filename;
  coordfile=opencoord(&grid_name,TRUE);
  if(coordfile==NULL)
    return EXIT_FAILURE;
  ngrid=numcoord(coordfile);
  if(cellsize_lon>0)
    res.lon=res.lat=cellsize_lon;
  else
  {
    getcellsizecoord(&cellsize_lon,&cellsize_lat,coordfile);
    res.lon=cellsize_lon;
    res.lat=cellsize_lat;
  }
  grid=newvec(Coord,numcoord(coordfile));
  if(grid==NULL)
  {
    printallocerr("grid");
    return EXIT_FAILURE;
  }
  for(i=0;i<numcoord(coordfile);i++)
    readcoord(coordfile,grid+i,&res);
  closecoord(coordfile);
  if(!israw)
  {
    if(ismeta)
      version=4;
    else
    {
      version=setversion;
      if(freadanyheader(file,&header,&swap,headername,&version,TRUE))
      {
        fprintf(stderr,"Error reading header of '%s'.\n",filename);
        free(grid);
        fclose(file);
        return EXIT_FAILURE;
      }
      if(version<3)
        header.datatype=LPJ_INT;
      if(version>CLM_MAX_VERSION)
      {
        fprintf(stderr,"Error: Unsupported version %d in '%s', must be less than %d.\n",
                version,filename,CLM_MAX_VERSION+1);
        free(grid);
        fclose(file);
        return EXIT_FAILURE;
      }
      filesize=getfilesizep(file)-headersize(headername,version);
      if(filesize!=(long long)header.nyear*header.ncell*header.nbands*typesizes[header.datatype])
        fprintf(stderr,"Warning: File size of '%s' does not match nbands*ncell*nyear.\n",filename);
    }
    if(header.nyear>1 || header.nstep>1)
    {
      fprintf(stderr,"No time axis set, but number of time steps>1 in '%s'.\n",
              filename);
      free(source);
      free(history);
      freeattrs(global_attrs,n_global);
      free(grid);
      fclose(file);
      return EXIT_FAILURE;
    }

    if(header.ncell!=ngrid)
    {
      fprintf(stderr,"Number of cells=%d in '%s' is different from %d in '%s'.\n",
              header.ncell,filename,ngrid,grid_filename);
      free(source);
      free(history);
      freeattrs(global_attrs,n_global);
      free(grid);
      fclose(file);
      return EXIT_FAILURE;
    }
    if(version==1 && cellsize_lon>0)
    {
      header.cellsize_lon=cellsize_lon;
      header.cellsize_lat=cellsize_lat;
    }
    if(fabs(header.cellsize_lon-res.lon)>min(res.lon,header.cellsize_lon)*0.5)
    {
      fprintf(stderr,"Longitudinal cell size=%g in '%s' differs from %g in '%s'.\n",
              header.cellsize_lon,filename,res.lon,grid_filename);
      free(source);
      free(history);
      freeattrs(global_attrs,n_global);
      free(grid);
      fclose(file);
      return EXIT_FAILURE;
    }
    if(fabs(header.cellsize_lat-res.lat)>min(res.lat,header.cellsize_lat)*0.5)
    {
      fprintf(stderr,"Latitudinal cell size=%g in '%s' differs from %g in '%s'.\n",
              header.cellsize_lat,filename,res.lat,grid_filename);
      free(source);
      free(history);
      freeattrs(global_attrs,n_global);
      free(grid);
      fclose(file);
      return EXIT_FAILURE;
    }
    if(ismeta)
    {
      header.cellsize_lon=res.lon;
      header.cellsize_lat=res.lat;
    }
  }
  else
  {
    header.cellsize_lon=(float)res.lon;
    header.cellsize_lat=(float)res.lat;
    header.datatype=LPJ_INT;
    header.ncell=ngrid;
    header.nbands=10;
    header.nyear=getfilesizep(file)/typesizes[header.datatype]/ngrid/header.nbands;
    if(getfilesizep(file) % (typesizes[header.datatype]*ngrid*header.nbands))
      fprintf(stderr,"Warning: file size of '%s' is not multiple of bands %d and number of cells %d.\n",filename,header.nbands,ngrid);
  }
  if(header.nbands!=10)
  {
    fprintf(stderr,"Number of bands=%d in '%s' must be 10.\n",
            header.nbands,filename);
    free(source);
    free(history);
    freeattrs(global_attrs,n_global);
    free(grid);
    fclose(file);
    return EXIT_FAILURE;
  }
  index=createindex(grid,ngrid,res,isglobal,FALSE);
  if(index==NULL)
  {
    free(source);
    free(history);
    freeattrs(global_attrs,n_global);
    free(grid);
    fclose(file);
    return EXIT_FAILURE;
  }
  free(grid);
  arglist=catstrvec(argv,argc);
  cdf=create_cdf(outname,source,history,miss,imiss,arglist,global_attrs,n_global,&header,compress,isnetcdf4,index);
  free(arglist);
  free(source);
  free(history);
  freeattrs(global_attrs,n_global);
  if(cdf==NULL)
  {
    freecoordarray(index);
    fclose(file);
    return EXIT_FAILURE;
  }
  rdata=newvec(Reservoir,ngrid);
  if(rdata==NULL)
  {
    printallocerr("data");
    fclose(file);
    close_cdf(cdf);
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid;i++)
  {
    if(readreservoir(rdata+i,swap,file))
    {
      fprintf(stderr,"Error reading reservoir data of cell %d.\n",i);
      free(rdata);
      fclose(file);
      close_cdf(cdf);
      return EXIT_FAILURE;
    }
    rdata[i].capacity*=1e-12; /* convert dm3 ->km3 */
  }
  fclose(file);
  rc=write_reservoir_cdf(cdf,rdata,ngrid,miss,imiss);
  free(rdata);
  close_cdf(cdf);
  return (rc) ? EXIT_FAILURE : EXIT_SUCCESS;
#else
  fprintf(stderr,"ERROR401: NetCDF is not supported in this version of %s.\n",strippath(argv[0]));
  return EXIT_FAILURE;
#endif
} /* of 'main' */
